File: app/src/main/java/su/linka/linkapaperboard/GridItemController.java
package su.linka.linkapaperboard;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import static su.linka.linkapaperboard.MainActivity.context;

public class GridItemController extends ArrayAdapter<String> {

    private int layoutResourceId;
    private String[] data;

    public GridItemController(@NonNull Context context, int resource, String[] arr) {
        super(context, resource, arr);
        layoutResourceId = resource;
        data = arr;
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        View row = convertView;

        if (row == null) {
            LayoutInflater inflater = context.getLayoutInflater();
            row = inflater.inflate(layoutResourceId, parent, false);
        }
        float size = GridController.getInstance().getGridHeight() / (GridController.getInstance().getSize() + 2);
        ((TextView) row).setTextSize(size / context.getResources().getDisplayMetrics().scaledDensity);

        String symbol = data[position];
        if(symbol!=null) {
            if (Cookie.getInstance().getUppercaseChecked()) symbol = symbol.toUpperCase();
            ((TextView) row).setText(symbol);
        }
        return row;
    }
}

File: app/src/main/java/su/linka/linkapaperboard/GridItemController.kt
package su.linka.linkapaperboard

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView
import java.util.Locale

class GridItemController(
    context: Context,
    private val layoutResourceId: Int,
    private val data: Array<String>,
    private val cellSizeProvider: () -> Float,
    private val uppercaseProvider: () -> Boolean
) : ArrayAdapter<String>(context, layoutResourceId, data) {

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val row = convertView ?: LayoutInflater.from(context).inflate(layoutResourceId, parent, false)
        val textView = row as TextView
        val size = cellSizeProvider()
        textView.textSize = size / context.resources.displayMetrics.scaledDensity

        var symbol = data[position]
        if (uppercaseProvider()) {
            symbol = symbol.uppercase(Locale.getDefault())
        }
        textView.text = symbol
        return row
    }
}
